### ISSUE TITLE
encode bytes directly

### ISSUE THREAD

**Issue by tsengalb99** [huggingface/tokenizers#1877]
Is there a way to directly encode bytes with a bpe based HF tokenizer without having to decode the string first?

**Comment by stephantul** [comment]
Hello, 

yes there is! You need a mapping from characters to bytes, and skip any pre-tokenization and normalization steps. 

```python
from tokenizers import Tokenizer

def bytes_to_unicode() -> dict[int, str]:
    """Converts byte values to Unicode characters for byte-level tokenization."""
    input_ids = list(range(33, 127)) + list(range(161, 173)) + list(range(174, 256))
    output_ids = input_ids[:]
    n = 0
    for char_ord in range(256):
        if char_ord not in input_ids:
            input_ids.append(char_ord)
            output_ids.append(256 + n)
            n += 1
    output_chars = [chr(c) for c in output_ids]
    return dict(zip(input_ids, output_chars, strict=True))

mapping = bytes_to_unicode()
tokenizer = Tokenizer.from_pretrained("gpt2")
tokenizer.normalizer = None
tokenizer.pre_tokenizer = None

byte_string = [65, 66, 233]
pseudo_string = "".join(mapping[b] for b in byte_string)

encoded = tokenizer.encode(pseudo_string, add_special_tokens=False)
```

Hope this helps!

**Comment by ArthurZucker** [comment]
In `transformers` we use this:
```python
def bytes_to_unicode():
    """
    Returns list of utf-8 byte and a mapping to unicode strings. We specifically avoids mapping to whitespace/control
    characters the bpe code barfs on.

    The reversible bpe codes work on unicode strings. This means you need a large # of unicode characters in your vocab
    if you want to avoid UNKs. When you're at something like a 10B token dataset you end up needing around 5K for
    decent coverage. This is a significant percentage of your normal, say, 32K bpe vocab. To avoid that, we want lookup
    tables between utf-8 bytes and unicode strings.
    """
    bs = (
        list(range(ord("!"), ord("~") + 1)) + list(range(ord("¡"), ord("¬") + 1)) + list(range(ord("®"), ord("ÿ") + 1))
    )
    cs = bs[:]
    n = 0
    for b in range(2**8):
        if b not in bs:
            bs.append(b)
            cs.append(2**8 + n)
            n += 1
    cs = [chr(n) for n in cs]
    return dict(zip(bs, cs))
```

then 
```python
       def token_bytes_to_string(b):
            return "".join([byte_encoder[ord(char)] for char in b.decode("latin-1")])
```

**Comment by arthur-ver** [comment]
Hi @ArthurZucker,

I was using the `ByteLevel` Pre-Tokenizer on byte strings that were decoded using Latin-1 to satisfy library's string input, and noticed some **undocumented behavior/bug**.

```python
from tokenizers import Tokenizer, pre_tokenizers
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer

tokenizer = Tokenizer(BPE())
tokenizer.pre_tokenizer = pre_tokenizers.ByteLevel(add_prefix_space=False)

trainer = BpeTrainer(vocab_size=1000, initial_alphabet=pre_tokenizers.ByteLevel().alphabet())

raw_data = [b"\xff\xff\xff\xff"] # includes raw byte that is illegal in UTF-8
iterator = [b.decode("latin-1") for b in raw_data] # Map directly to first 256 Unicode characters

tokenizer.train_from_iterator(iterator, trainer=trainer)

encoded = tokenizer.encode(raw_data[0].decode("latin-1"))

print([tokenizer.id_to_token(t) for t in encoded.ids])
# Output: ['Ã¿Ã¿Ã¿Ã¿']
# Expected: ['ÿÿÿÿ']
```

Instead of learning the single byte pattern for `b"\xff"` (which should be `"ÿ"`), the tokenizer learns to merge `"Ã"` and `"¿"`. The true raw byte is completely masked by the UTF-8 encoding boundary when the string is passed to Rust. This results in:

1) **Vocabulary Pollution:** 50% of all possible single bytes (anything `>= 0x80`) are artificially expanded into 2-byte UTF-8 sequences, wasting vocabulary slots.
2) **Dead Tokens:** The base characters for high-value bytes (like `"ÿ"`) will permanently have a count of 0 in the vocabulary, as the Rust pre-tokenizer mathematically cannot encounter illegal UTF-8 bytes like `0xFF`.

Proposed Solutions:
1) **Accept bytes directly:** Allow `train_from_iterator` and `tokenizer.encode()` to accept Python bytes objects. If bytes are provided and a ByteLevel pre-tokenizer is active, apply your proposed `bytes_to_unicode()` helper function.
2) **Documentation Warning:** Add a clear warning in the ByteLevel documentation explaining that passing arbitrary binary data decoded via latin-1 will result in vocabulary bloat due to UTF-8 conversion.

**Comment by stephantul** [comment]
You should skip the `ByteLevel` pretokenizer, this would give you what you want. The byte level pretokenizer is re-encoding the raw byte string again, but that's exactly the opposite of what you are trying to achieve.

**Comment by arthur-ver** [comment]
@stephantul Yes, I figured it out by now :)

However, the docs still remain unclear of the purpose and limitations of `ByteLevel` pre-tokenizer:
> This pre-tokenizer takes care of replacing all bytes of the given string with a corresponding representation, as well as splitting into words.

**Comment by ArthurZucker** [comment]
happy to review a PR with a clearer doc @arthur-ver 😉

**Comment by michaelalbada** [comment]
Hey @ArthurZucker, I took the initiative to address this here: https://github.com/huggingface/tokenizers/pull/2195
